test/cpp/end2end/grpclb_end2end_test.cc

```diff
@@ -521,17 +540,12 @@ TEST_F(SingleBalancerTest, Vanilla) {
       0);
   // Make sure that trying to connect works without a call.
   channel_->GetState(true /* try_to_connect */);
-  // Send 100 RPCs per server.
-  const auto& statuses_and_responses =
-      SendRpc(kMessage_, kNumRpcsPerAddress * num_backends_);
-
-  for (const auto& status_and_response : statuses_and_responses) {
-    const Status& status = status_and_response.first;
-    const EchoResponse& response = status_and_response.second;
-    EXPECT_TRUE(status.ok()) << "code=" << status.error_code()
-                             << " message=" << status.error_message();
-    EXPECT_EQ(response.message(), kMessage_);
-  }
+
+  // We need to wait for all backends to come online.
+  for (size_t i = 0; i < num_backends_; ++i) WaitForBackend(i);
```

Comment:
I don't quite understand why this can be more efficient. 

For the time side: the total waiting time is almost equal to the time for the slowest backend to become ready in both conditions.

For the space side: the total number of sent RPCs is like _the total waiting time * RPC sending rate_, which should also be the same in both conditions.

What am I missing?

Reply:
The approach I suggested minimizes the total number of RPCs sent.

There are two loops here.  The outer loop iterates over the backends in order.  For each backend in the outer loop, we invoke the inner loop, which keeps sending RPCs until the current backend from the outer loop receives a request.  Any time the inner loop sends more than one RPC, we have sent those excess RPCs to backends other than the current backend from the outer loop.  However, we then reset the counters before continuing the outer loop, which means that we forget that those other backends have already seen a request, so we will always send at least one RPC (and possibly more) in the inner loop for each subsequent iteration of the outer loop.

Let's take a very minimal example with only 2 backends, indexes 0 and 1, and that index 1 becomes ready first.  The current code will do the following:

- outer loop: try index 0
- (backend 1 becomes ready)
- inner loop: send RPC, which goes to backend 1
- (backend 0 becomes ready)
- inner loop: send RPC, which goes to backend 0
- we got a request for backend 0, so reset counters for both backends and break out of inner loop
- outer loop: try index 1
- inner loop: send RPC, which goes to index 1
- we got a request for backend 1, so reset counters for both backends and break out of inner loop

Note that the above sent 3 RPCs.  The main problem here is that we reset counters for all backends between iterations of the outer loop.  This means that after the first iteration of the outer loop, we forget that backend 1 has already seen a request, which is why we need to send another RPC for the second iteration of the outer loop.

My proposed change is to avoid resetting the counters until after all iterations of the outer loop, which avoids this problem.  We would thus send only 2 RPCs in the example above.

Reply:
Ah.. yes. It does reduce the number of sent RPCs because it only resets the counters after all the backends are ready. Thanks for the example.